### DISCUSSION TITLE
Using Python, how can I copy all widths of a source font?

### DISCUSSION THREAD

**Discussion by sean256** [fontforge/fontforge#4654] [Q&A]
I got a source font that has `thin`, `regular`, `bold`, etc, all bundled up in a ttf file.

Following the simple script example in the docs I can get some glyphs copied but only one width is in the outputted font.

```python
import fontforge

sourceFont=fontforge.open("./someFont.ttf")

sourceFont.selection.select(("ranges", None), "A", "Z")
sourceFont.copy()

newFont=fontforge.font()
newFont.paste()

newFont.fontname = "Test Font"
newFont.generate("outputFont.ttf")
```

With the above, `outputFont.ttf` will only have one width and it will be named "book". How do I copy all of them?

**Comment by ctrlcctrlv** [comment]
Variable fonts aren't well supported. FontForge isn't the right tool to manipulate them.

**Reply by sean256** [reply]
Oh darn. Do you know of another tool with scripting ability that can do this?

**Reply by ctrlcctrlv** [reply]
You're sure it's a variable font, right, and not simply a TrueType Collection?

**Reply by ctrlcctrlv** [reply]
Well anyway, the problem you seem to be trying to solve is the problem of font subsetting.

If this really is a variable font, I would use fontmake along with the method described in googlefonts/fontmake#343 to subset a variable UFO font. If your variable font isn't available as UFO sources anywhere, you can use the Python ufo-extractor library to make one from a TTF. Here's an example script: https://github.com/googlefonts/robotoslab/blob/main/scripts/bin/ttf2ufo

If it's simply a TrueType collection, while there might be a way to do it with FontForge, what I would do is, use AFDKO's [`otc2otf`](https://github.com/adobe-type-tools/afdko/blob/develop/python/afdko/otc2otf.py). That will split the TrueType Collection, then you can subset the generated fonts (with e.g. FontForge), and only then use `otf2otc` to recreate the TrueType Collection. Although these programs are called `otf`, they support `ttf` `glyf` table I'm pretty sure.
